sql/src/main/java/com/vmware/ddlog/ir/DDlogProgram.java

```diff
@@ -36,33 +36,51 @@
     public final List<DDlogTypeDef> typedefs;
     public final List<DDlogFunction> functions;
     public final List<DDlogRelationDeclaration> relations;
+    public final List<DDlogIndexDeclaration> indexes;
     public final List<DDlogRule> rules;
     public final List<DDlogImport> imports;
 
-    DDlogProgram(List<DDlogTypeDef> typedefs, List<DDlogFunction> functions, List<DDlogRelationDeclaration> relations, List<DDlogRule> rules, List<DDlogImport> imports) {
+    DDlogProgram(List<DDlogTypeDef> typedefs, List<DDlogFunction> functions, List<DDlogRelationDeclaration> relations,
+                 List<DDlogIndexDeclaration> indexes,
+                 List<DDlogRule> rules, List<DDlogImport> imports) {
         super(null);
         this.typedefs = typedefs;
         this.functions = functions;
         this.relations = relations;
+        this.indexes = indexes;
         this.rules = rules;
         this.imports = imports;
     }
 
     public DDlogProgram() {
         this(new ArrayList<DDlogTypeDef>(), new ArrayList<DDlogFunction>(),
-                new ArrayList<DDlogRelationDeclaration>(), new ArrayList<DDlogRule>(),
+                new ArrayList<DDlogRelationDeclaration>(),
+                new ArrayList<DDlogIndexDeclaration>(),
+                new ArrayList<DDlogRule>(),
                 new ArrayList<DDlogImport>());
     }
 
     @Override
     public String toString() {
-        String[] parts = new String[5];
-        parts[0] = String.join("\n", Linq.map(this.imports, DDlogImport::toString)) + "\n";
-        parts[1] = String.join("\n", Linq.map(this.typedefs, DDlogTypeDef::toString));
-        parts[2] = String.join("\n", Linq.map(this.functions, DDlogFunction::toString));
-        parts[3] = String.join("\n", Linq.map(this.relations, DDlogRelationDeclaration::toString));
-        parts[4] = String.join("\n", Linq.map(this.rules, DDlogRule::toString));
-        return String.join("\n", parts);
+        // We do this to preserve the correctness of existing tests, which are sensitive to white-space
+        if (this.indexes.size() == 0) {
+            String[] parts = new String[5];
```

Comment:
This starts to look ugly. Perhaps just make a list and keep adding to it?

Reply:
Take a look at the update? No matter what, if we want to avoid changing the results of many of the tests, it will look somewhat ugly (i.e. check for whether `this.indexes.size() == 0`.

Reply:
Was arguing against allocating exactly 6/5 elements